hera_qm/xrfi.py

```diff
@@ -222,6 +224,223 @@ def robust_divide(num, den):
+    if len(eigenval_cutoff) == 1 and len(filter_centers) > 1:
+        eigenval_cutoff = len(filter_centers) * eigenval_cutoff
+
+    if wgts is None:
+        wgts = np.ones_like(data)
+
+    else:
+        wgts = deepcopy(wgts)
+
+    # Compute first-pass model and residuals
+    model, _, _ = dspec.fourier_filter(
+        freqs, data, wgts, filter_centers, filter_half_widths, mode=mode,
+        suppression_factors=suppression_factors, eigenval_cutoff=eigenval_cutoff,
+        cache=cache
+    )
+    res = data - model
+
+    # Compute modified z-score
+    sigma = np.median(
+        np.abs(res - np.median(res, keepdims=True, axis=1)), axis=1, keepdims=True,
+    )
+    sigma /= 0.675
+    wgts = np.where(np.abs(res) < nsig * sigma, 1, 0)
+
+    # Recompute model with first pass weights
+    model, _, _ = dspec.fourier_filter(
+        freqs, data, wgts, filter_centers, filter_half_widths, mode=mode,
+        suppression_factors=suppression_factors, eigenval_cutoff=eigenval_cutoff,
+        cache=cache
+    )
+    res = data - model
+
+    # Use noise standard deviation given or estimate it from residuals
+    if noise is not None:
+        sigma = noise
+
+    else:
+        sigma = np.median(
+            np.abs(res - np.median(res, keepdims=True, axis=1)), axis=1, keepdims=True,
+        )
+        sigma /= 0.675
+
+    # Determine weights
+    wgts = np.where(np.abs(res) <= nsig * sigma, 1, 0)
+    return wgts
+
+
+def channel_diff_flagger(data, noise, nsig=10, kernel_widths=[1, 2, 4], wgts=None):
+    """
+    Flags auto-correlations using channel differencing kernels
+
+    Parameters:
+    ----------
+    data: np.ndarray
+        2D data array of the shape (time, frequency)
+    noise: np.ndarray
+        2D array for containing an estimate of the noise standard deviation. Must be the same shape as the data
+    nsig: float, default=10
+        The number of sigma in the metric above which to flag pixels.
+    kernel_width: list
+        Widths of the convolution kernels used to produce model
+    wgts: np.ndarray, default=None
+        2D array of real weights. Must be the same shape as data.
+
+    Returns:
+    -------
+    wgts: np.ndarray
+        Array containing weights for each antenna with values of either zero or one with
+        zero identifying channels found to have RFI
+    """
+    # If weights is None, set value to one
+    if wgts is None:
+        wgts = np.ones_like(data)
+
+    # Iterate through differencing kernels of different widths
+    for width in kernel_widths:
+        kernel = np.ones(2 * width + 1)
+        kernel_std = np.sqrt(1 + 1 / (2 * width))
+        dw = data * wgts
+        residual = np.zeros_like(data)
+
+        # For each time integration, identify outliers
+        for t in range(data.shape[0]):
+            _d = np.convolve(dw[t], kernel, mode="same") - dw[t]
+            _w = np.convolve(wgts[t], kernel, mode="same") - wgts[t]
+            residual[t] = dw[t] - _d / _w.clip(1, np.inf)
```

Comment:
What is the clip doing and what do its bounds mean?

Reply:
The clip is preventing divide by zero errors. The `_w` array shouldn't have values less than one except at the location of flagged channels, which should have zeros.

Reply:
I still don't get this. Why? Can't one cook up a set of wgts where that wouldn't be true? 

It seems like maybe this function is written assuming that all weights are 1 or zero, but there's no guarantee of that from the docstring or any kind of error handling.

Reply:
I replaced this with flags instead of weights. I also removed the clipping and went with a robust_divide instead.